torch/csrc/jit/codegen/cuda/index_compute.cpp

```diff
@@ -501,7 +501,10 @@ void IndexCompute::handle(Split* split) {
     index_map_[in_id] = ir_builder.create<kir::Int>(0);
     extent_map_[in_id] = ir_builder.create<kir::Int>(0);
   } else if (zero_merged_in && outer_zero) {
+    // Hack to propagate vectorization correctly. Otherwise extent of vectorized
+    // dim will get lost
     index_map_[in_id] = inner_ind;
+    index_map_[in_id] = ir_builder.addExpr(inner_ind, ir_builder.zeroVal());
```

Comment:
I understand the intention, but I really don't like this hack.

It seems that what we need here is to maintain a set of domains whose indices are "zero", that is, domains that only consist of domains that are left of CA positions. The problem is that we try to infer the set from index values that are `isZeroInt`, which may also include domains that are right of CA positions because of vectorization and broadcast. 

This ambiguity can be worked around by the zero addition hack, but it's too hacky and implicit to my eyes. I'd prefer to make the logic more explicit, if possible. For example, it seems to me that it could be also avoided by maintaining the set directly. There's `zero_merged_in_` ; we could also have something like `zero_domains_`. It would be populated by `indexMapFromTV`, and then updated along `split` and `merge`. With `split`, only when both outer and inner are in `zero_domains_`, the `in` domain would be also added to `zero_domains_`. With `merge`, when the out domain is `zero_domains_`, both the two inputs would be also added to `zero_domains_`. With this, the vector extent would not be lost since the input to split would be only added to `zero_merged_in_` but not to `zero_domains_`, so it should avoid take the path at line 500.

Reply:
I'll see if I could improve later. Filed #1039